File: core/src/adapters/arm_pmu.rs
use crate::adapters::EnergyAdapter;
use crate::EnergyResult;
use hardware_plugins::{HardwarePlugin, Measurement};

pub struct ArmPmuAdapter {
    base: super::BaseAdapter,
}

impl ArmPmuAdapter {
    pub fn new(plugin: Box<dyn HardwarePlugin>) -> Self {
        Self {
            base: super::BaseAdapter::new("arm_pmu", plugin),
        }
    }
}

impl EnergyAdapter for ArmPmuAdapter {
    fn name(&self) -> &str {
        self.base.name()
    }

    fn initialize(&mut self) -> EnergyResult<()> {
        Ok(())
    }

    fn shutdown(&mut self) -> EnergyResult<()> {
        Ok(())
    }

    fn read_measurements(&self) -> EnergyResult<Vec<Measurement>> {
        let measurement = self.base.plugin().get_measurement()?;
        Ok(vec![measurement])
    }
} 

File: src/measurement/src/adapters/arm_pmu.cpp
#include "plugin/hardware_plugin.hpp"

namespace codegreen {

class ARMPMUAdapter : public HardwarePlugin {
public:
    ARMPMUAdapter() = default;
    ~ARMPMUAdapter() override = default;

    std::string name() const override { return "ARM PMU"; }
    std::unique_ptr<Measurement> get_measurement() const override { return nullptr; }
    bool init() override { return true; }
    void cleanup() override {}
    bool is_available() const override { return false; }
};

} // namespace codegreen
